restapi/src/main/java/io/stargate/web/docsapi/service/query/search/resolver/CnfResolver.java

```diff
@@ -0,0 +1,232 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.web.docsapi.service.query.search.resolver;
+
+import com.bpodgursky.jbool_expressions.And;
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Literal;
+import com.bpodgursky.jbool_expressions.options.ExprOptions;
+import com.bpodgursky.jbool_expressions.rules.Rule;
+import com.bpodgursky.jbool_expressions.rules.RuleList;
+import com.bpodgursky.jbool_expressions.rules.RulesHelper;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import io.reactivex.rxjava3.core.Flowable;
+import io.stargate.web.docsapi.dao.DocumentDB;
+import io.stargate.web.docsapi.dao.Paginator;
+import io.stargate.web.docsapi.service.DocsApiConfiguration;
+import io.stargate.web.docsapi.service.RawDocument;
+import io.stargate.web.docsapi.service.query.FilterExpression;
+import io.stargate.web.docsapi.service.query.FilterPath;
+import io.stargate.web.docsapi.service.query.rules.TrueFilterExpressions;
+import io.stargate.web.docsapi.service.query.search.weigth.ExpressionWeightResolver;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Resolver that accepts the {@link And} expression in the CNF form.
+ */
+public class CnfResolver implements DocumentsResolver {
+
+  private final Expression<FilterExpression> expression;
+
+  private final ExpressionWeightResolver<FilterExpression> weightResolver;
+
+  private final DocumentsResolver parent;
+
+  public CnfResolver(
+          Expression<FilterExpression> expression,
+          ExpressionWeightResolver<FilterExpression> weightResolver, DocumentsResolver parent) {
+    if (!And.EXPR_TYPE.equals(expression.getExprType())) {
+      throw new IllegalArgumentException(
+          "CNF resolver expect the AND expression in the conjunctive normal form.");
+    }
+
+    this.expression = expression;
+    this.weightResolver = weightResolver;
+    this.parent = parent;
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p>
+   *
+   */
+  @Override
+  public Flowable<RawDocument> getDocuments(
+      DocumentDB db,
+      DocsApiConfiguration configuration,
+      String keyspace,
+      String collection,
+      Paginator paginator) {
+
+    // resolve the next and pipe
+    DocumentsResolver next = this.resolveNext();
+    return next.getDocuments(db, configuration, keyspace, collection, paginator);
```

Comment:
Again, would it be possible to do all the resolution under the `resolve` method and let `getDocuments` handle only execution? I'd also rename `getDocuments` to `execute` in that case since the "get" part is not really immediate. WDYT?

Reply:
Let's shortly discuss this..

Reply:
Done.